File: anwendung/src/test/java/domaene/HaushaltsbuchführungBeginnenSteps.java
package domaene;

import com.github.haschi.haushaltsbuch.api.ereignis.ImmutableHauptbuchWurdeAngelegt;
import com.github.haschi.haushaltsbuch.api.ereignis.ImmutableHaushaltsbuchAngelegt;
import com.github.haschi.haushaltsbuch.api.ereignis.ImmutableJournalWurdeAngelegt;
import com.github.haschi.haushaltsbuch.api.kommando.ImmutableBeginneHaushaltsbuchführung;
import cucumber.api.java.de.Dann;
import cucumber.api.java.de.Wenn;
import org.axonframework.commandhandling.gateway.CommandGateway;
import testsupport.DieWelt;

import javax.inject.Inject;
import javax.inject.Singleton;
import java.util.UUID;

import static org.assertj.core.api.Assertions.assertThat;

@Singleton
public final class HaushaltsbuchführungBeginnenSteps
{
    private final CommandGateway commandGateway;

    private final DieWelt welt;

    @Inject
    public HaushaltsbuchführungBeginnenSteps(final CommandGateway commandGateway, final DieWelt welt)
    {
        this.commandGateway = commandGateway;
        this.welt = welt;
    }

    @Wenn("^ich mit der Haushaltsbuchführung beginne$")
    public void ich_mit_der_Haushaltsbuchführung_beginne()
    {
        final UUID haushaltsbuchId = UUID.randomUUID();
        this.commandGateway.sendAndWait(ImmutableBeginneHaushaltsbuchführung.builder().id(haushaltsbuchId).build());
        this.welt.setAktuelleHaushaltsbuchId(haushaltsbuchId);
    }

    @Dann("^werde ich ein neues Haushaltsbuch angelegt haben")
    public void dann_werde_ich_ein_neues_haushaltsbuch_angelegt_haben()
    {
        assertThat(this.welt.aktuellerEreignisstrom()).contains(ImmutableHaushaltsbuchAngelegt.builder()
                                                                        .id(this.welt.getAktuelleHaushaltsbuchId())
                                                                        .build());
    }

    @Dann("^werde ich ein Hauptbuch mit Kontenrahmen zum Haushaltsbuch angelegt haben")
    public void dann_werde_ich_ein_hauptbuch_mit_kontenrahmen_zum_haushaltsbuch_anlegen()
    {

        assertThat(this.welt.aktuellerEreignisstrom()).contains(ImmutableHauptbuchWurdeAngelegt.builder()
                                                                        .haushaltsbuchId(this.welt
                                                                                                 .getAktuelleHaushaltsbuchId())
                                                                        .build());
    }

    @Dann("^werde ich ein Journal zum Haushaltsbuch angelegt haben")
    public void dann_werde_ich_ein_journal_zum_haushaltsbuch_anlegen()
    {
        assertThat(this.welt.aktuellerEreignisstrom()).contains(ImmutableJournalWurdeAngelegt.builder()
                                                                        .aktuelleHaushaltsbuchId(this.welt
                                                                                                         .getAktuelleHaushaltsbuchId())
                                                                        .build());
    }
}

File: testit/src/test/kotlin/com/github/haschi/domain/haushaltsbuch/HaushaltsbuchführungBeginnenSteps.kt
package com.github.haschi.domain.haushaltsbuch

import com.github.haschi.domain.haushaltsbuch.testing.DieWelt
import com.github.haschi.domain.haushaltsbuch.testing.Kontozeile
import com.github.haschi.dominium.haushaltsbuch.core.model.commands.BeginneHaushaltsbuchführung
import com.github.haschi.dominium.haushaltsbuch.core.model.events.EröffnungsbilanzkontoErstellt
import com.github.haschi.dominium.haushaltsbuch.core.model.queries.LeseInventar
import com.github.haschi.dominium.haushaltsbuch.core.model.values.Aggregatkennung
import com.github.haschi.dominium.haushaltsbuch.core.model.values.Buchung
import com.github.haschi.dominium.haushaltsbuch.core.model.values.Eröffnungsbilanzkonto
import com.github.haschi.dominium.haushaltsbuch.core.model.values.Inventar
import cucumber.api.java.de.Dann
import cucumber.api.java.de.Wenn
import org.assertj.core.api.Assertions.assertThat

class HaushaltsbuchführungBeginnenSteps(private val welt: DieWelt)
{
    @Wenn("^ich die Haushaltsbuchführung beginne$")
    fun ichDieHaushaltsbuchführungBeginne()
    {
        val inventar = welt.query.query(LeseInventar(welt.aktuelleInventur),
                Inventar::class.java).get()

        welt.aktuellesHaushaltsbuch = welt.haushaltsbuchführung.send(
                BeginneHaushaltsbuchführung(
                        id = Aggregatkennung.neu(),
                        inventar = inventar)).get()
    }

    @Dann("^werde ich folgendes Eröffnungsbilanzkonto im Hauptbuch erstellt haben:$")
    fun `dann werde ich folgendes Eröffnungsbilanzkonto im Hauptbuch Erstellt haben`(
            eröffnungsbilanzkonto: List<Kontozeile>)
    {
        val eröffnungsbilanzkontoErstellt =
                welt.historie.bezüglich(welt.aktuellesHaushaltsbuch)
                        .filter { it.payload is EröffnungsbilanzkontoErstellt }
                        .map { m -> m.payload as EröffnungsbilanzkontoErstellt }
                        .first()

        assertThat(eröffnungsbilanzkontoErstellt.eröffnungsbilanzkonto)
                .isEqualTo(Eröffnungsbilanzkonto(
                        haben = eröffnungsbilanzkonto
                                .map { it.haben ?: Buchung.leer }
                                .filter { it != Buchung.leer }
                                .toList(),
                        soll = eröffnungsbilanzkonto
                                .map { it.soll ?: Buchung.leer }
                                .filter { it != Buchung.leer }
                                .toList()))
    }
}
